File: ai_guardrail/aerorisk/models/event.py
"""
Event models for the AeroRisk AI pipeline.
"""

from enum import Enum
from pydantic import BaseModel, Field


class OrderSide(str, Enum):
    BUY = "BUY"
    SELL = "SELL"


class OrderAction(str, Enum):
    SUBMIT = "SUBMIT"
    CANCEL = "CANCEL"
    FILL = "FILL"
    MODIFY = "MODIFY"


class EventType(str, Enum):
    ORDER = "ORDER"
    TRADE = "TRADE"
    RISK_DECISION = "RISK_DECISION"
    CIRCUIT_BREAKER = "CIRCUIT_BREAKER"
    AUDIT = "AUDIT"


class OrderEvent(BaseModel):
    """Order event from the matching engine."""

    order_id: str
    account_id: str
    symbol: str
    side: OrderSide
    action: OrderAction
    price: int  # Fixed-point (price * 10^6)
    quantity: int
    timestamp: float
    client_order_id: str | None = None
    user_agent: str | None = None
    ip_address: str | None = None

    class Config:
        use_enum_values = True


class TradeEvent(BaseModel):
    """Trade execution event."""

    trade_id: str
    order_id: str
    counterparty_order_id: str
    symbol: str
    price: int
    quantity: int
    buyer_account_id: str
    seller_account_id: str
    timestamp: float
    aggressor_side: OrderSide


class RiskDecisionEvent(BaseModel):
    """Risk decision from AI pipeline."""

    decision_id: str
    order_id: str | None = None
    account_id: str
    decision: str  # ALLOW, FLAG, BLOCK, ADJUST_LIMIT
    risk_score: float
    reasons: list[str] = Field(default_factory=list)
    sanctions_match: bool = False
    anomaly_flags: list[str] = Field(default_factory=list)
    timestamp: float
    agent_version: str = "v1.0.0"


class CircuitBreakerEvent(BaseModel):
    """Circuit breaker state change."""

    breaker_id: str
    account_id: str | None = None
    symbol: str | None = None
    state: str  # CLOSED, OPEN, HALF_OPEN
    trigger_reason: str
    triggered_at: float
    expires_at: float | None = None


class AuditEvent(BaseModel):
    """Audit log event."""

    audit_id: str
    event_type: EventType
    entity_id: str
    action: str
    actor: str
    details: dict
    timestamp: float
    checksum: str | None = None

File: engine/pkg/event/event.go
package event

import "time"

type Event struct {
ID        uint64    `json:"id"`
Type      EventType `json:"type"`
Timestamp time.Time `json:"timestamp"`
Payload   []byte    `json:"payload"`
}

func NewEvent(id uint64, t EventType, payload []byte) *Event {
return &Event{
ID:        id,
Type:      t,
Timestamp: time.Now(),
Payload:   payload,
}
}

type OrderEvent struct {
OrderID   string `json:"order_id"`
AccountID string `json:"account_id"`
Symbol    string `json:"symbol"`
Side      string `json:"side"`
Price     int64  `json:"price"`
Quantity  int64  `json:"quantity"`
Timestamp int64  `json:"timestamp"`
}

type TradeEvent struct {
TradeID    string `json:"trade_id"`
OrderID    string `json:"order_id"`
MatchID    string `json:"match_id"`
Symbol     string `json:"symbol"`
Price      int64  `json:"price"`
Quantity   int64  `json:"quantity"`
BuyerID    string `json:"buyer_id"`
SellerID   string `json:"seller_id"`
Timestamp  int64  `json:"timestamp"`
}

type RiskEvent struct {
EventID      string `json:"event_id"`
AccountID    string `json:"account_id"`
OrderID      string `json:"order_id,omitempty"`
Decision     string `json:"decision"`
RiskScore    float64 `json:"risk_score"`
Reason       string `json:"reason"`
CircuitBreak bool   `json:"circuit_break"`
Timestamp    int64  `json:"timestamp"`
}

type AuditEvent struct {
EventID   string `json:"event_id"`
UserID    string `json:"user_id"`
Action    string `json:"action"`
Resource  string `json:"resource"`
Details   string `json:"details,omitempty"`
Timestamp int64  `json:"timestamp"`
}
